### PR TITLE
Made Diggy scenario config generic

### PR DESCRIPTION
**Pull request by linaori**
With this change, you only have to point at a namespace and it can load feature from there.

Theoretical config that @Jayefuu could use for biter line battles:

```lua
-- config/biter_line_battles.lua
return {
    features = {
        -- enables different teams
        team_setup = {
            enabled = true,
        },
        -- enables a custom world builder 
        world_builder = {
            enabled = true,
            layout_preset = 'north-to-south',
        },
    },
}
```

```lua
-- map_layout.lua
require 'utils.scenario'('map_gen.preset.biter_line_battles', require 'config/biter_line_battles')
-- would load
--  * map_gen.preset.biter_line_battles.team_setup.lua
--  * map_gen.preset.biter_line_battles.world_builder.lua
```

Might have to figure out how to nicely combine this with the generic config from #502

### CODE REVIEW

File: utils/scenario.lua [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+require 'utils.table'
+require 'utils.core'
+
+local format = string.format
+
+global.scenario_loaded = false
+
+return function (scenario_namespace, config)
+    if global.scenario_loaded then
+        error(format('Cannot register '%s' scenario multiple times.', scenario_namespace))
```
[the remark this answers is not in the document]
Comment (mheguy) [reply utils/scenario.lua]: Not sure why you marked this resolved, the bot is correct.
image

**Review by mheguy — DISMISSED** [review]

**Comment by mheguy** [comment]
This change seems antithetical to a lot of things we're building towards.

I think we want things to be simple for mapmakers since they might not be the most proficient coders. Creating a situation where they have to write (in your example) 13 lines of code and then create a much more complex `require` in map_layout to achieve the end-goal of requiring 2 files seems like a significant hurdle.

This also seems to encourage more map-specific functionality as opposed to building in a way where other maps can take advantage of features by enhancing the whole `namespace` concept.

Lastly, this makes `map_layout` itself much more daunting to a would-be contributor first looking at it and especially so for an end-user trying to find the map they want to play.

Could you explain the rationale behind this?

File: utils/scenario.lua (hunk 1 above)
[the remark this answers is not in the document]
Comment (linaori) [reply utils/scenario.lua]: Hm odd, I'll check it when I'm home.

**Comment by linaori** [comment]
> This change seems antithetical to a lot of things we're building towards

One of the things we're working towards, is easy config, which I'm hoping to do with this change.

> I think we want things to be simple for mapmakers since they might not be the most proficient coders.

That's why I'm attempting to abstract away some layers behind configuration and convention.

> This also seems to encourage more map-specific functionality as opposed to building in a way where other maps can take advantage of features by enhancing the whole namespace concept.

They are for scenario specific features, not the RedMew generic (features namespace). the idea behind it is that it's easy for scenario makers to isolate a piece of (experimental) code while not having the overhead of run-time configuration.

> Lastly, this makes map_layout itself much more daunting to a would-be contributor first looking at it and especially so for an end-user trying to find the map they want to play.

Ideally I'd only ever see people use config files, so this would not be an issue for me. In this example, I've inlined the `diggy.lua` in here, but I could revert this change and have `map_gen/combined/diggy.lua` do this particular loading. Meaning map_layout would be unchanged by this.

> Could you explain the rationale behind this?

I'm trying to find a way where it's easier for scenarios to be added with pre-defined boundaries, like a framework. Jayefuu wants to build a scenario as well and while this could easily go in a single file, it would miss out on easy configuration and probably leak a lot of stuff in and out of globals. I took Diggy as example, as it has a clear structure: 
 - A feature can be enabled. If not, it's not loaded at all, reducing overhead, dependencies, commands etc from being registered
 - There's no fiddling with requires. You just add a file.lua in your directory and add the same name to your config
 - Config is injected into the register function, meaning you don't have to fiddle with globals

With this small framework, I'm hoping to eventually make it easier for people to _create_ a scenario, but _also_ to configure one if they have limited lua knowledge.

File: utils/scenario.lua (hunk 1 above)
[the remark this answers is not in the document]
Comment (linaori) [reply utils/scenario.lua]: This should now be fixed, I forgot to remove the quotes transforming this to use `string.format`.

**Comment by linaori** [comment]
Summary of changes, as it turned out a bit bigger because of a merge:
 - SetupPlayer no longer has cheats, but uses the cheats introduced in the `create_player.lua` feature
 - `map_layout.lua` has been restored to its simplicity
 - Fixed the feedback as provided by Herr Stickler

**Comment by mheguy** [comment]
I appreciate your explanation. It does make more sense now. I think currently my biggest sticking point is:

>There's no fiddling with requires. You just add a file.lua in your directory and add the same name to your config

I feel like this encourages the creation of scenario-specific modules. There are absolutely some cases where it's necessary, but as much as possible it would be nice if they were made in a way that is accessible to other scenarios.

To cite a specific example: crashsite's `weapon_balance.lua` concept is cool. You can modify the bonuses granted by weapon damage research. But because it's tucked away inside the crash folder nobody would think to use it. (As it happens, for the rotten apples map recently we wanted a lot of the same functionality so ended up copy+pasting essentially the whole of weapon_balance into the rotten_apples map preset)

tl;dr Make scenario-specific modules opt-in, not opt-out. Please don't make it the default or easier than standard modules.

**Comment by linaori** [comment]
> I feel like this encourages the creation of scenario-specific modules. There are absolutely some cases where it's necessary, but as much as possible it would be nice if they were made in a way that is accessible to other scenarios.

I agree, this is what I'm aiming for most of the time. However, sometimes it's really scenario specific, like pretty much all "features" in Diggy. Perhaps features is not the correct term, and it should be called a "module" instead? I also think that we should aim to make stuff generic when it makes sense and review accordingly.

Regarding the above, I want to try and extract the market GUI for the crash site to be usable outside it. This would give us the benefit of adding biter images for biter lane wars for example. I honestly don't have problems with code evolving from something scenario specific to generic. One of the examples could be the alien script that I have to calculate which aliens should spawn at which evolution with what %.

**Comment by mheguy** [comment]
I think we both understand features and modules to be interchangeable words. We both understand each other's meaning.

So for example something like the room generator or cave collapse: of course they are and should be diggy specific.

I'm referring more to modules like `SetupPlayer`, 'NightTime', and `AlienSpawner`

>  I honestly don't have problems with code evolving from something scenario specific to generic.

I think that's where the conflict arises. In my experience, when people start creating just for themselves and not others it perpetuates down the line. As in my example of weapon balance, I copy pasted the contents of the entire file and tweaked 1-2 lines rather than refactoring it into something generic and using that.

Crash site and diggy are intimidating because of how many unique modules they have. Fundamentally, I don't know or understand their interdependencies and so I don't trust myself to properly extract features that are unique to them even if they are amazing candidates for generic modules. And so, as in the case of `player_create`, rather than using the great code you had written a month and a half earlier, another similar feature was created.

**Comment by linaori** [comment]
Leaving this for now,
